### PR TITLE
gpioioctl:Implement ioctl access to Linux GPIO chips/lines.

### PR DESCRIPTION
**Pull request by gsexton**
Fixes host issue #39

First, let me start by saying I tend to work in lots of languages, and my fluency in golang is not really high. If you see non-idiomatic things you can't live with, just let me know. Similarly, I know you're the maintainer, and it's your call in implementations.

This is a pretty big add. If you feel like a call would be beneficial, I can do a voice call.

## Terminology

I'm using the term "Line", and not "Pin". Line does implement conn/v3/gpio.PinIn, PinOut, and PinIO. This is kind of a difference. Line is an attribute of the chip, while pin is an attribute of the board. It seems to me that we want the code to work across boards, regardless of how a board physically presents a GPIO chipset's line. So a pin is a physical element that
the chip's line is connected to. I can see potentially a GPIO line that's not exposed as a physical pin on a board. As an aside, the ioctl()s use the Line terminology. However, if you think that's unnecessarily confusing, I can change it.

## Tests
Refer to Tests.md

### consumer

In the ioctl(), there's a "consumer" field. The kernel docs say:

```
    a functional name for the consumer of this GPIO line as set by whatever is using it,
    will be empty if there is no current user but may also be empty if the consumer doesn’t
    set this up
```

I'm writing the program_name@pid into this field. I can see other programs use it too. For example, I see PPS bound to a GPIO Chip on my Pi. You might consider adding 

func Consumer() string

to the gpio.Pin interface for the next revision.

### PinFunc

I really don't understand the PinFunc interface, so I didn't implement it. I looked it over, and from what I could tell, I didn't like the idea, but perhaps I didn't understand it. If you can give me a better idea of the intent, I'll give it a go.

### PWM

PWM isn't really part of the GPIO ioctl User Space API. The kernel docs say:
```
Do NOT abuse userspace APIs to control hardware that has proper kernel drivers.
There may already be a driver for your use case, and an existing kernel driver
is sure to provide a superior solution to bitbashing from userspace.
```
I've stubbed in a PWM method to satisfy PinOut, but it should probably be deprecated from the PinOut interface. Here's the what the doc says:
```
pwm-gpio: drivers/pwm/pwm-gpio.c is used to toggle a GPIO with a high resolution
timer producing a PWM waveform on the GPIO line, as well as Linux high resolution
timers can do.
```
## MultiPin IO
Multi-Pin IO is handled by the new construct LineSet. The ioctl() interface makes this pretty straight forward. Single line input/output are identical to multi-line. Other than configuration, no clever tricks are required. You just call Read() and it
returns the pin values as a uint64 bitmap. Essentially, a LineSet provides methods for reading or writing all lines in one go. LineSetLine provides convenience methods to allow writing individual pins and implements PinIn, PinOut, and PinIO.  Additionally, functions of pins can be different within the set. Some can be input, and others output.

A common use case for a LineSet might be doing a HD44780 LCD display which has 4 or 8 lines + a strobe line, and a backlight line.

### CODE REVIEW

File: gpioioctl/Testing.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# Testing
```
Comment (maruel) [inline gpioioctl/Testing.md:1]: This should go in https://github.com/periph/cmd/tree/main/periph-smoketest/gpiosmoketest

**Comment by maruel** [comment]
I haven't looked at the code yet but answering a few questions for sake of speed:
- PinFunc is mostly relevant for MCU provided hardware pins, so it's okay to ignore for now.
- PWM: sounds good. It's okay to just return an error.
- Line vs Pin and multi-pins: it's something I wanted to create a conn/v4 for but lost steam sadly as the refactor became big; I want to make all stateful functions like PWM() to accept a context.Context.

File: gpioioctl/Testing.md (hunk 1 above)
Comment (gsexton) [reply gpioioctl/Testing.md]: I've done the restructure. The sysfs stuff is still conflicting. Do you understand it?

File: gpioioctl/basic_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+func TestChips(t *testing.T) {
+	if len(Chips) <= 0 {
+		t.Fatal("Chips contains no entries.")
+	}
+	chip := Chips[0]
+	if len(chip.Name()) == 0 {
+		t.Error("chip.Name() is 0 length")
+	}
+	if len(chip.Path())==0 {
```
Comment (maruel) [inline gpioioctl/basic_test.go:31]: You may want to configure gofmt -w -s and/or goimports -w automatically in your IDE. The github checks won't pass otherwise.

File: gpioioctl/basic_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+)
+
+var _test_line *GPIOLine
+
+func init() {
+	_, _ = driverreg.Init()
+}
+
+func TestChips(t *testing.T) {
+	if len(Chips) <= 0 {
```
Comment (maruel) [inline gpioioctl/basic_test.go:24]: Can this test pass on macOS and Windows? Options could be to mock or move this kind of testing to https://github.com/periph/cmd/pull/7.

File: gpioioctl/basic_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+// periph.io/x/cmd/v3/periph-smoketest/gpiosmoketest
+// folder.
+import (
+	"testing"
+
+	"periph.io/x/conn/v3/driver/driverreg"
+	"periph.io/x/conn/v3/gpio/gpioreg"
+)
+
+var _test_line *GPIOLine
```
Comment (maruel) [inline gpioioctl/basic_test.go:17]: FYI, go naming convention would be testLine.

File: gpioioctl/basic_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+func init() {
+	_, _ = driverreg.Init()
```
Comment (maruel) [inline gpioioctl/basic_test.go:20]: Would you mind checking for errors here? Otherwise it could create hard-to-diagnose failures.

File: gpioioctl/example_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	// Test by flashing an LED.
+	led := gpioreg.ByName("GPIO5")
+	fmt.Println("Flashing LED ", led.Name())
+	for i := range 20 {
+		_ = led.Out((i % 2) == 0)
+		time.Sleep(500 * time.Millisecond)
+	}
+	_ = led.Out(true)
+
+	testRotary(chip, "GPIO20", "GPIO21", "GPIO19")
```
Comment (maruel) [inline gpioioctl/example_test.go:36]: Caveat: Maybe, I'm not sure.
You may want to inline the function to make it look better. You can confirm that it looks good:
```
go install golang.org/x/pkgsite/cmd/pkgsite@latest
pkgsite
```
then navigate to http://localhost:8080/periph.io/x/host/v3

File: sysfs/sysfssmoketest/benchmark.go [hunk 7]
```diff
@@ -1,69 +0,0 @@
-// Copyright 2017 The Periph Authors. All rights reserved.
```
Comment (maruel) [inline sysfs/sysfssmoketest/benchmark.go:1]: sysfs/sysfssmoketest/benchmark.go is a remnant where I failed (forgot!) to move it to cmd. I'm sorry about that. I'd prefer to do this as a separate PR since this PR is already fairly big.

File: gpioioctl/lineset.go [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+func (lsl *LineSetLine) WaitForEdge(timeout time.Duration) bool {
+	return false
+}
+
+// Pull returns the configured PullUp/PullDown value for this line.
+func (lsl *LineSetLine) Pull() gpio.Pull {
+	return lsl.pull
+}
+
+// DefaultPull - return gpio.PullNoChange. Reviewing the GPIO v2 Kernel
```
Comment (maruel) [inline gpioioctl/lineset.go:391]: Suggestion:

// DefaultPull return gpio.PullNoChange.
//
// The GPIO v2 Kernel doesn't support this.

**Review by maruel — COMMENTED** [review]
Awesome. Don't be afraid by the number of comments, it's a chunk change. Thanks again.

File: gpioioctl/gpio.go [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+
+// The consumer name to use for line requests. Initialized in init()
+var consumer []byte
+
+// The set of GPIO Chips found on the running device.
+var Chips []*GPIOChip
+
+type Label string
+
+var DirectionLabels = []Label{"NotSet", "Input", "Output"}
```
[the remark this answers is not in the document]
Comment (gsexton) [reply gpioioctl/gpio.go:54]: I didn't originally, but the smoke test references the values, so it's an artifact of moving functional tests out of the package.

File: gpioioctl/ioctl.go [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+)
+
+type gpiochip_info struct {
+	name  [_GPIO_MAX_NAME_SIZE]byte
+	label [_GPIO_MAX_NAME_SIZE]byte
+	lines uint32
+}
+
+func (ci *gpiochip_info) String() string {
+	return fmt.Sprintf("{\"Name\": \"%s\", \"Label\": \"%s\", \"Lines\": %d}", string(ci.name[:]), string(ci.label[:]), ci.lines)
```
[the remark this answers is not in the document]
Comment (gsexton) [reply gpioioctl/ioctl.go:86]: I've escaped the JSON. I tend to do a lot of back-end work where debuggers just aren't a thing. Debugging information is what I can get in the logs, and so I tend to be pretty complete. From a support standpoint, having all of this dump in a legible format is just super-handy. 

OTOH, there's no reasonable use case for the String() to return end user displayed data here. In this case, the end users are developers. I thought about using yaml, but had already written the JSON.

File: sysfs/sysfssmoketest/benchmark.go (hunk 7 above)
Comment (gsexton) [reply sysfs/sysfssmoketest/benchmark.go:1]: restored.

File: gpioioctl/example_test.go (hunk 6 above)
Comment (gsexton) [reply gpioioctl/example_test.go:36]: I've fixed this and it shows in the format as expected.

File: gpioioctl/basic_test.go (hunk 3 above)
Comment (gsexton) [reply gpioioctl/basic_test.go:24]: Can we do this with build tags and conditional compilation? Do you specify any --tags args in the go test command?

File: gpioioctl/basic_test.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
+package gpioioctl
```
Comment (maruel) [inline gpioioctl/basic_test.go:1]: The usual Go style is:

```
// Copyright 2024 The Periph Authors. All rights reserved.
// Use of this source code is governed under the Apache License, Version 2.0
// that can be found in the LICENSE file.

// Basic tests. More complete test is contained in the
// periph.io/x/cmd/periph-smoketest/gpiosmoketest
// folder.

package gpioioctl

import (
```

Note that cmd is not versioned, since it's a set of executables, not an API.

File: gpioioctl/basic_test.go (hunk 5 above)
Comment (maruel) [reply gpioioctl/basic_test.go:20]: Did you forget to push a new version? It's okay to just panic.

File: gpioioctl/basic_test.go (hunk 3 above)
Comment (maruel) [reply gpioioctl/basic_test.go:24]: Even so, the test assumes the host will expose chips on linux. This will not be the case on VMs.

Comment (gsexton) [reply gpioioctl/basic_test.go:24]: OK, I've added a mock to make this work regardless if init() finds any gpio chips.

File: gpioioctl/basic_test.go (hunk 5 above)
Comment (gsexton) [reply gpioioctl/basic_test.go:20]: Sorry, I somehow didn't commit this change.

File: gpioioctl/gpio.go [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+		if len(charBytes) >= _GPIO_MAX_NAME_SIZE {
+			charBytes = charBytes[:_GPIO_MAX_NAME_SIZE-1]
+		}
+		consumer = charBytes
+
+		driverreg.MustRegister(&drvGPIO)
+	}
+}
+
+func escapeJSONString(str string) string {
```
Comment (maruel) [inline gpioioctl/gpio.go:598]: FYI, the better implementation is to use:
```
d, _ := json.Marshal(str)
return string(d)
```

https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/encoding/json/encode.go;l=160 ultimately leads to https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/encoding/json/encode.go;l=967 which has many subtleties.

That sad, I don't think you need this function given the other comment.

**Review by maruel — COMMENTED** [review]
Notes about JSON.
I changed the GitHub settings so the actions should start immediately upon push, so you don't need to wait for me to click the Approve button.

File: gpioioctl/gpio.go [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+		log.Println(err)
+		return false
+	}
+	return data.bits&0x01 == 0x01
+}
+
+// String returns information about the line in JSON format.
+func (line *GPIOLine) String() string {
+
+	return fmt.Sprintf("{\"Line\": %d, \"Name\": \"%s\", \"Consumer\": \"%s\", \"Direction\": \"%s\", \"Pull\": \"%s\", \"Edges\": \"%s\"}",
```
[the remark this answers is not in the document]
Comment (gsexton) [reply gpioioctl/gpio.go]: Implemented

File: gpioioctl/ioctl.go [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+		if li.flags&FLAG_LIST[i] == FLAG_LIST[i] {
+			flag_str += FLAG_NAMES[i][0] + " "
+		} else {
+			if len(FLAG_NAMES[i][1]) > 0 {
+				flag_str += FLAG_NAMES[i][1] + " "
+			}
+		}
+	}
+	return fmt.Sprintf("{\"Name\": \"%s\", \"Consumer\": \"%s\", \"Offset\": %d, \"# Attrs\": %d, \"Flags\": \"%s\" 0x%x, \"Attributes\": \"%s\"}",
```
[the remark this answers is not in the document]
Comment (maruel) [reply gpioioctl/ioctl.go:184]: Did you mean to push a new version?

[the remark this answers is not in the document]
Comment (gsexton) [reply gpioioctl/ioctl.go:184]: Can you help me out with what I'm doing wrong? I look at the complete file for the PR, and it doesn't show the lines you're quoting.

[the remark this answers is not in the document]
Comment (maruel) [reply gpioioctl/ioctl.go:184]: Here's what I see with the URL you pasted:
Capture d’écran du 2024-09-09 13-25-27

Here's what I'm commenting on:
Capture d’écran du 2024-09-09 13-27-41

**Comment by maruel** [comment]
I'll merge once the checks are green.

**Comment by gsexton** [comment]
@maruel So i'm pretty sure the reason that netlink was not loading in host init() is because the pipeline blocked infinitely trying receive the message from the netlink socket. I got creative and added a 2s timeout to the onewire.Init(). That seems to stop the pipeline from timing out. It now passes the Linux pipeline build. I tested it on a pi with a couple of DS18B20 units attached and it worked as expected.

I've fixed all of the lint issues in gpioioctl, sysfs, and netlink, which is code that I've touched.

The rest is all gosec G115 possible overflow errors. I don't think they're false positives. I'd recommend disabling that check in the pipeline.

File: host_linux.go [hunk 15]
```diff
@@ -5,6 +5,8 @@
 package host
 
 import (
-	// Make sure sysfs drivers are registered.
+	// Make sure required drivers are registered.
+	_ "periph.io/x/host/v3/gpioioctl"
+	_ "periph.io/x/host/v3/netlink"
```
Comment (maruel) [inline host_linux.go:10]: Let's do one change at a time. This change is already very large. Let's do netlink modification in a separate PR. Thanks.

Comment (gsexton) [reply host_linux.go]: OK, I've reverted those files. I'll re-open the issue I had earlier and make a new PR for it.

**Comment by gsexton** [comment]
@maruel Were you waiting on me for something more? I think I've got everything you asked for completed.

**Comment by maruel** [comment]
Yes sorry I just needed to silence gosec, done in https://github.com/periph/host/commit/1e513a0cdf6a479e05b38984a467787868c7d006.

**Comment by maruel** [comment]
Thanks a lot! It's really appreciated.
